Cancel Google sheet form upload when activity destroyed

lakshyagupta21 opened the pull request:
Closes #1777 

#### What has been done to verify that this works as intended?
Tested on physical device (Android 5.0 and 6.0) and behaviour found correct

#### Why is this the best possible solution? Were any other approaches considered?
This approach stops uploading forms to Google Sheets whenever activity gets destroyed and takes back the user to 'Send Finalized Form' Screen. Sometimes Android may kill activity in the background due to memory pressure. There may be other approaches possible.
 
#### Are there any risks to merging this code? If so, what are they?
No
#### Do we need any specific form for testing your changes? If so, please attach one.
I tested using  [odk-sheets.zip](https://github.com/opendatakit/collect/files/1693219/odk-sheets.zip)

lakshyagupta21 committed:
stop upload when activity destroyed

lakshyagupta21 committed:
Implementation method changed

grzesiek2010 commented:
@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "needs information"

grzesiek2010 reviewed (changes requested):
I think we could just add finish() in onDestroy() and that's all only one line. Am I wrong?

lakshyagupta21 commented:
@grzesiek2010  Yeah this will also serve our purpose and we need to cancel the upload process in onDestroy() otherwise it will keep on uploading it. 
I'll update the PR if you want.

grzesiek2010 commented:
> and we need to cancel the upload process in onDestroy() otherwise it will keep on uploading it

I think we don't. The problem was that the `GoogleSheetsUploaderActivity` activity was recreated. If we use finish everything should be ok. Please test it.

lakshyagupta21 commented:
If we use `finish()` it won't cancel the task by itself which is uploading the forms. If we don't cancel it and if user again 'Send' the forms then it will create duplicates on Google Sheets.

grzesiek2010 commented:
> If we don't cancel it and if user again 'Send' the forms then it will create duplicates on Google Sheets.

I'm not able to notice such a behavior. Did you test it?

lakshyagupta21 commented:
Yes, if we keep onDestroy method like
```
 @Override
    protected void onDestroy() {
        if (instanceGoogleSheetsUploader != null) {
            instanceGoogleSheetsUploader.setUploaderListener(null);
        }
        finish();
        super.onDestroy();
    }
```
And if we again open the app you'll see the list gets updated whenever any forms gets uploaded. And in the meantime if user again selects some forms and send then the task will again get executed for the selected form. This will lead to duplicate submissions for the forms.

grzesiek2010 commented:
Ok so please use only that `cancel` and `finish` since it works and it's simpler.

grzesiek2010 commented:
@opendatakit-bot unlabel "needs information"
@opendatakit-bot label "needs testing"

lognaturel commented:
It's not clear to me what the state is here. @lakshyagupta21 said:

> And if we again open the app you'll see the list gets updated whenever any forms gets uploaded. And in the meantime if user again selects some forms and send then the task will again get executed for the selected form. This will lead to duplicate submissions for the forms.

This is a problem you are able to reproduce with the code you now have in this PR, right?

lakshyagupta21 commented:
> This is a problem you are able to reproduce with the code you now have in this PR, right?

@lognaturel now current code solves this problem too. Actually, @grzesiek2010 asked me to test things by just placing finish() in onDestroy() but it creates a new problem adding task cancelling block solves the problem.

mmarciniak90 commented:
@lognaturel @lakshyagupta21 
I'm still able to double the submission
![1831](https://user-images.githubusercontent.com/3907554/36309416-67fd523e-131c-11e8-843f-fa887a1bd5cd.gif)

RESULT:
![1831](https://user-images.githubusercontent.com/3907554/36309424-6cf81666-131c-11e8-9d10-817aa26f23a8.png)

It is easier to reproduce when image is added probably because sending the form takes longer. 

@opendatakit-bot unlabel "needs testing"

lakshyagupta21 committed:
Conditional stmts added to cancel the task

lakshyagupta21 commented:
@mmarciniak90 I've changed some code, earlier if you try even if you cancel the task the form gets uploaded to the sheets now I have added some conditional statements before each major part of upload activity to check whether the task is cancelled or not, earlier it was checked only once.
@grzesiek2010 Please review the code.

grzesiek2010 commented:
@lakshyagupta21 @lognaturel @mmarciniak90 

There is a big refactoring pending for testing #1851 and here there is a couple of checks like this one https://github.com/opendatakit/collect/pull/1831/files#diff-5e0801a16659617bd27be6ed6903b142R210

I would don't even know where it should be added to the new version of `InstanceGoogleSheetsUploader` so let's test and merge #1851 first and then we can add this one (maybe we won't need as many checks then?)

Does anyone disagree?

lognaturel commented:
That's a very good point, @grzesiek2010, thank you. Agreed it makes sense to prioritize verifying #1851 and merging it as soon as possible and then reevaluating this.

grzesiek2010 commented:
@lakshyagupta21 #1851 has been merged so you can go ahead.

lakshyagupta21 commented:
@grzesiek2010 I think I committed some mistake while merging the code, because of this so many files got changed. Can you help me with this?

lakshyagupta21 committed:
String resource added

grzesiek2010 commented:
@lakshyagupta21 could you fix conflicts?

mmarciniak90 commented:
**Tested with success!**

I was not able to create doubled submissions on Google Sheets with turned on `Do not keep activities` option

Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0, 8.0.

@opendatakit-bot  unlabel "needs testing"
@opendatakit-bot  label "behavior verified"

lognaturel commented:
Thanks all and especially for your patience, @lakshyagupta21!
